sklearn/ensemble/iforest.py

```diff
@@ -250,17 +250,28 @@ def decision_function(self, X):
         """
         # code structure from ForestClassifier/predict_proba
         # Check data
-        X = self.estimators_[0]._validate_X_predict(X, check_input=True)
+        X = check_array(X, accept_sparse='csr')
         n_samples = X.shape[0]
 
         n_samples_leaf = np.zeros((n_samples, self.n_estimators), order="f")
         depths = np.zeros((n_samples, self.n_estimators), order="f")
 
-        for i, tree in enumerate(self.estimators_):
-            leaves_index = tree.apply(X)
-            node_indicator = tree.decision_path(X)
+        if self._max_features == X.shape[1]:
+            subsample_features = False
+        else:
+            subsample_features = True
+
+        for i, (tree, features) in enumerate(zip(self.estimators_,
+                                                 self.estimators_features_)):
+            if subsample_features:
+                X_subset = X[:, features]
+            else:
+                X_subset = X
+            leaves_index = tree.apply(X_subset)
+            node_indicator = tree.decision_path(X_subset)
```

Comment:
It seems a bit strange that we do this inference twice. But I know that's not the present issue.